### ISSUE TITLE
[Bug]: Protocol error when calling puppeteer.connect()

### ISSUE THREAD

**Issue by drmrbrewer** [puppeteer/puppeteer#8944]
### Bug description

I am using the basic approach as set out [in this post][1] to connect from a client docker container to any one of a number of chrome docker containers (in a docker swarm/service, potentially across several servers behind nginx, deployed using [CapRover][2]).

In each chrome container I maintain a pool (just a simple array) of `browser` objects, and direct incoming requests to an appropriate `browser` as follows (very similar to the linked post):

	import http from 'node:http';       // https://nodejs.org/api/http.html
	import httpProxy from 'http-proxy'; // https://www.npmjs.com/package/http-proxy

	const proxy = new httpProxy.createProxyServer({ ws: true });

	// an array (pool) of pre-launched and managed browser objects...
	const browsers = [ ... ];

	http
	  .createServer()
	  .on('upgrade', (req, socket, head) => {
		  const browser = browsers[Math.floor(Math.random() * browsers.length)]; // in reality I don't just pick a browser at random
		  const target = browser.wsEndpoint();
		  proxy.ws(req, socket, head, { target });
	  })
	  .listen(3222);

The above is listening at `ws://srv-captain--chrome:3222` (communication is "internal" over the docker network between containers).

Then, in my client container, I connect to the common endpoint `ws://srv-captain--chrome:3222` as follows:

	import puppeteer from 'puppeteer'; // https://www.npmjs.com/package/puppeteer (using version 17.1.3 at time of posting this)
	try {
		const browser = await puppeteer.connect({ browserWSEndpoint: 'ws://srv-captain--chrome:3222' });
	} catch (err) {
		console.error('error connecting to browser', err);
	}

This works really well, except that I am getting *occasional/inconsistent* errors like these when calling `puppeteer.connect()` in the client container above:

    Protocol error (Emulation.setDeviceMetricsOverride): Session closed. Most likely the page has been closed.
    Protocol error (Performance.enable): Target closed.

Almost always, if I simply try to connect again, the connection is made without further error, and at the first attempt.

I have no idea why the error is complaining that the `page has been closed` or `Target closed` since, at this point in the process, I'm not attempting to interact with any `page`, and I know from listening for `browser.on('disconnected'...)`, and also monitoring the chromium processes themselves, that each `browser` in the array is still working fine... none has crashed.

Any idea what's going on here?

__________

**UPDATE after further testing**

Of course, in the client container we don't connect to a `browser` just for the sake of it, like in the above snippet, but to open a `page` and do some stuff with the `page`.  In practice, in the client container it's more like the following test snippet:

	const doIteration = function (i) {
		return new Promise(async (resolve, reject) => {
			// mimic incoming requests coming in at random times over a short period by introducing a random initial delay...
			await new Promise(resolve => setTimeout(resolve, Math.random() * 5000));
			// now actually connect...
			let browser;
			try {
				browser = await puppeteer.connect({ browserWSEndpoint: `ws://srv-captain--chrome:3222?queryParam=loop_${i}` });
			} catch (err) {
				reject(err);
				return;
			}
			// now that we have a browser, open a new page...
			const page = await browser.newPage();
			// do something useful with the page (not shown here) and then close it..
			await page.close();
			// now disconnect (but don't close) the browser...
			browser.disconnect();
			resolve();
		});            
	};

	const promises = [];
	for (let i = 0; i < 15; i++) {
		promises.push( doIteration(i) );
	}

	try {
		await Promise.all(promises);
	} catch (err) {
		console.error(`error doing stuff`, err);
	}

Each iteration above is being performed multiple times concurrently... I am using `Promise.all()` on an array of iteration promises to mimic multiple concurrent incoming requests in my production code.  The above is enough to reproduce the problem... the error doesn't happen on calling `puppeteer.connect()` with *every* iteration, just some.

So there seems to be some sort of interplay between opening/closing a `page` in one iteration, and calling `puppeteer.connect()` in another, despite closing the `page` and disconnecting the `browser` properly in each iteration?  This probably also explains the `Most likely the page has been closed` error message when calling `puppeteer.connect()` if there is some hangover relating to a `page` closed in another iteration... though for some reason this error occurs when calling `puppeteer.connect()`?

__________

With the use of a pool of `browser` objects in the `browsers` array, and a docker swarm having multiple containers on multiple servers, each `upgrade` message could be received at a different container (which could even be on a different server) and could be routed to a different `browser` in the `browsers` array.  But I now think that this is a red herring, because in the further testing I narrowed the problem down by routing all requests to `browsers[0]` and also scaling the service down to just one container... so that the `upgrade` messages are always handled by the same container on the same server and routed to the same `browser`... and the problem still occurs.
__________

Full stacktrace for the above-mentioned error:

	Error: Protocol error (Emulation.setDeviceMetricsOverride): Session closed. Most likely the page has been closed.
		at CDPSession.send (file:///root/workspace/myclientapp/node_modules/puppeteer/lib/esm/puppeteer/common/Connection.js:281:35)
		at EmulationManager.emulateViewport (file:///root/workspace/myclientapp/node_modules/puppeteer/lib/esm/puppeteer/common/EmulationManager.js:33:73)
		at Page.setViewport (file:///root/workspace/myclientapp/node_modules/puppeteer/lib/esm/puppeteer/common/Page.js:1776:93)
		at Function._create (file:///root/workspace/myclientapp/node_modules/puppeteer/lib/esm/puppeteer/common/Page.js:242:24)
		at runMicrotasks (<anonymous>)
		at processTicksAndRejections (node:internal/process/task_queues:96:5)
		at async Target.page (file:///root/workspace/myclientapp/node_modules/puppeteer/lib/esm/puppeteer/common/Target.js:123:23)
		at async Promise.all (index 0)
		at async BrowserContext.pages (file:///root/workspace/myclientapp/node_modules/puppeteer/lib/esm/puppeteer/common/Browser.js:577:23)
		at async Promise.all (index 0)

  [1]: https://stackoverflow.com/a/49090055/4070848
  [2]: https://caprover.com/

### Puppeteer version

17.1.3

### Node.js version

16.17.0

### npm version

8.15.0

### What operating system are you seeing the problem on?

Linux

**Comment by OrKoN** [comment]
Thanks for so many details! Do you know if it re-produces if you run without the docker? I.e., just starting chromium with the --remote-debugging-port and connecting using the script you provided in the update?

**Comment by drmrbrewer** [comment]
@OrKoN I'll see if I can reproduce this without docker, and get back with any info.

**Comment by drmrbrewer** [comment]
@OrKoN it seems to reproduce with a local chromium executable too... see sample below.  Note that it's entirely possible that I'm misunderstanding how the concept of connecting to and then disconnecting from a `browser` is meant to work, and this may be why a `page` I open (and then close) in one iteration is conflicting with something in another iteration.  I had assumed/hoped that I can open/close pages independently in each iteration (though on the same `browser` endpoint) without any issues.  So it may be something I've misunderstood, rather than a bug... but if it's a misunderstanding then it would be good to know how I'm meant to be connecting to a `browser` (local or remote) like I'm trying to do!  Thanks!

	const launchOptions = {
		args: [ '--no-sandbox', '--disable-setuid-sandbox' ],
		headless: true
	};
	const browser = await puppeteer.launch(launchOptions);
	const target = browser.wsEndpoint();

	const doStuff = function (i) {
		return new Promise(async (resolve, reject) => {
			// as before, mimic incoming requests at overlapping times...
			await MyFunctions.setTimeoutPromise(Math.random() * 5000);

			let browser;
			try {
				browser = await puppeteer.connect({ browserWSEndpoint: target });

			} catch (err) {
				// errors out here frequently...
				console.error(`error connecting to browser loop ${i}`, err);
				reject(err);
				return;
			}

			try {
				const page = await browser.newPage();
				// in this test, just close page without doing anything with it...
				await page.close();
				// disconnect (but don't close) browser...
				browser.disconnect();

			} catch (err) {
				console.error(`error using browser loop ${i}`, err);
				reject(err);
				return;
			}

			resolve();
		});            
	};

	const promises = [];
	for (let i = 0; i < 25; i++) {
		promises.push( doStuff(i) );
	}

	try {
		await Promise.all(promises);
	} catch (err) {
		console.error(`error doing stuff`, err);
	}

**Comment by OrKoN** [comment]
It does look like a bug, I will try to reproduce it, thanks for the script!

**Comment by OrKoN** [comment]
@drmrbrewer yeah it looks plausible, thanks for bisecting

**Comment by drmrbrewer** [comment]
I can also confirm that when I revert to version `15.5.0` (where we connect to a remote browser instance as per the original question) I am no longer seeing these particular errors when calling `puppeteer.connect()`.

However, I am still seeing occasional errors thrown in my code for a call which happens immediately *after* connecting... this is again easily reproducible by adding one line to call `await browser.pages()`, as follows:

    const pages = await browser.pages(); // <-- NEW LINE
    const page = await browser.newPage();

again it complains about `Target closed`... but not every time.  Do you know why this might be?  It does seem to throw the error in one loop when calling `browser.pages()` at the same time that I'm closing a page in another loop... is this really expected from `browser.pages()`?  In my code, I need to check the number of pages before deciding whether or not to call `browser.newPage()` yet (I have a page limit on each browser).

**Comment by OrKoN** [comment]
@drmrbrewer the fundamental problem is that the pages in the browser instance are a shared state and you open/close them in parallel. So if you call `browser.pages()` from one session, it gets all pages includes the ones created by the other connections. Since the browser is not a database, it can only be eventually consistent when handling targets, and thus we get a problem that while one connection is in the middle of initialising a page created by another connection, that other connection closes the page causing errors in the first connection. Of course, we should not throw on this and suppress errors but fundamentally you will still see flakiness using `browser.pages()`: it might be returning the different number of pages depending on which pages happened to be active.

**Comment by OrKoN** [comment]
We have regressed here a bit since 15.5.0 as we used `browser.pages` inside `.connect` https://github.com/puppeteer/puppeteer/blob/main/src/common/BrowserConnector.ts#L134 Once we fix it you should be able to achieve better isolation using https://pptr.dev/next/api/puppeteer.browser.createincognitobrowsercontext but you should not call `browser.pages()` in your program still.

**Comment by drmrbrewer** [comment]
> Of course, we should not throw on this and suppress errors but fundamentally you will still see flakiness using browser.pages()

I can deal with some flakiness (in my case I don't mind if I open 16 pages on a browser, or stop at 14, when I set the limit at  15... mainly I want to avoid opening 30 pages).  I can understand the reasons for the issue, but ultimately I would still expect this function to return a (promise which resolves to) an array of pages, whether this is the pages just before something else started messing with the pages state, or waiting until the pages state becomes stable again and then return it.  Better than just rejecting with an error?  I guess I should keep a separate count of pages in the browser, but the danger in this (for me) is that this might get out of sync with the actual number of pages, unless I'm very very careful to catch all possible places where the page count can increment or decrement... I just feel more secure calling `browser.pages()`.

**Comment by drmrbrewer** [comment]
Anyway, thanks so much for your quick response to this issue.  By reverting to `15.5.0` for now, and being more aware of vulnerabilities associated in general with `browser.pages()` (*), I've managed to get my production system running clean as a whistle again.  Looking forward to the fixes in the next version, hopefully!

(*) for now I'm just polling `browser.pages()` until I get a non-error response (or give up trying)... I'm hopeful that `browser.pages()` can be changed to do something along these lines, internally?

**Comment by drmrbrewer** [comment]
@OrKoN thanks again for this upcoming fix!  Regarding the problem with using `browser.pages()` when concurrent requests are involved, is this likely to be something that can be improved?  Shall I raise a separate issue for this?

**Comment by OrKoN** [comment]
@drmrbrewer calling `browser.pages()` should not fail anymore but the number of pages returned will depend on how many pages are actually open in the browser at the moment. I don't think there is anything to be done about it except for using the `browserContext.pages()` and creating a new browsing context for a concurrent connection.

**Comment by drmrbrewer** [comment]
> calling `browser.pages()` should not fail anymore but the number of pages returned will depend on how many pages are actually open in the browser at the moment

This is absolutely good enough for my purposes... definitely better to return something sensible rather than fail... thanks.
